adapter/prometheus2/prometheus.go

```diff
@@ -0,0 +1,289 @@
+		once     sync.Once
+	}
+
+	prom struct {
+		metrics map[string]prometheus.Collector
+		kinds   map[string]config.Params_MetricInfo_Kind
+	}
+)
+
+var (
+	charReplacer = strings.NewReplacer("/", "_", ".", "_", " ", "_")
+
+	_ metric.HandlerBuilder = &factory{}
+	_ metric.Handler        = &prom{}
+)
+
+// GetBuilderInfo returns the BuilderInfo associated with this adapter implementation.
+func GetBuilderInfo() adapter.BuilderInfo {
+	return adapter.BuilderInfo{
+		Name:        "prometheus",
+		Description: "Publishes prometheus metrics",
+		SupportedTemplates: []string{
+			metric.TemplateName,
+		},
+		CreateHandlerBuilder: func() adapter.HandlerBuilder { return newFactory(newServer(defaultAddr)) },
+		DefaultConfig:        &config.Params{},
+		ValidateConfig:       func(msg adapter.Config) *adapter.ConfigErrors { return nil },
+	}
+}
+
+func newFactory(srv server) *factory {
+	return &factory{srv, prometheus.NewPedanticRegistry(), sync.Once{}}
+}
+
+func (f *factory) ConfigureMetricHandler(map[string]*metric.Type) error {
+	return nil
+}
+
+func (f *factory) Build(c adapter.Config, env adapter.Env) (adapter.Handler, error) {
+	var serverErr error
+	f.once.Do(func() {
+		serverErr = f.srv.Start(env, promhttp.HandlerFor(f.registry, promhttp.HandlerOpts{}))
+	})
+	if serverErr != nil {
+		return nil, fmt.Errorf("could not start prometheus server: %v", serverErr)
+	}
+
+	cfg := c.(*config.Params)
+	var metricErr *multierror.Error
+	metricsMap := make(map[string]prometheus.Collector, len(cfg.Metrics))
+	kinds := make(map[string]config.Params_MetricInfo_Kind, len(cfg.Metrics))
+	for _, m := range cfg.Metrics {
+		switch m.Kind {
+		case config.GAUGE:
+			c, err := registerOrGet(f.registry, newGaugeVec(m.Name, m.Description, m.LabelNames))
+			if err != nil {
+				metricErr = multierror.Append(metricErr, fmt.Errorf("could not register metric: %v", err))
+				continue
+			}
+			metricsMap[m.Name] = c
+			kinds[m.Name] = config.GAUGE
+		case config.COUNTER:
+			c, err := registerOrGet(f.registry, newCounterVec(m.Name, m.Description, m.LabelNames))
+			if err != nil {
+				metricErr = multierror.Append(metricErr, fmt.Errorf("could not register metric: %v", err))
+				continue
+			}
+			metricsMap[m.Name] = c
+			kinds[m.Name] = config.COUNTER
+		case config.DISTRIBUTION:
+			c, err := registerOrGet(f.registry, newHistogramVec(m.Name, m.Description, m.LabelNames, m.Buckets))
+			if err != nil {
+				metricErr = multierror.Append(metricErr, fmt.Errorf("could not register metric: %v", err))
+				continue
+			}
+			metricsMap[m.Name] = c
+			kinds[m.Name] = config.DISTRIBUTION
+		default:
+			metricErr = multierror.Append(metricErr, fmt.Errorf("unknown metric kind (%d); could not register metric", m.Kind))
+		}
+	}
+	return &prom{metricsMap, kinds}, metricErr.ErrorOrNil()
+}
+
+// TODO: figure out ownership semantics, because factory.close will never be called in the new scheme.
+func (f *factory) Close() error {
+	return f.srv.Close()
```

Comment:
The problem is that server is a singleton, while handler instances are not. You cannot bind multiple servers to the same port, so you cannot just move it into the handler. Even the unit tests break doing this.